### DISCUSSION TITLE
Acquire List of All Event Types

### DISCUSSION THREAD

**Discussion by SE2Dev** [biodiv/contactjs#29] [General]
#28 can be solved by augmenting the `HTMLElement` event listener types to add ContactJS specific overloads which would accept a `GestureEvent` instead of a regular `Event`.

Do we have a way to deterministically get a list of all event type combinations, or do we simply need to keep track manually?

e.g.

```
pan
panend
pinchstart
pinch
pinchend
twofingerpanstart
twofingerpan
twofingerpanend
...
```

Most likely we'd also want a list of all possible directions as well, since I believe the following are also valid event types:

```
panleft
panright
...
```

**Comment by biodiv** [comment]
Would it make sense to add a list of possible events to each gesture class, or would this make no sense as it is currently not used in the logic?

Currently there are the following gestures and events:

### Tap
Recognized after the pointer has been removed from the surface emitting the `tap` event. No directional events, no beginning, no end and thus no `tapend` or `tapstart`.

### Press
Press has a beginning and an end. `pressstart`, `pressend`, `press` events exist. No directional events.

### Pan
Obviously, `panstart`, `panend`, `pan`, `panup`, `pandown`, `panleft`, `panright` exist. The `Pan` class also emits `swipe`, `swipeleft`, `swiperight`, `swipeup`, `swipedown`. A swipe requires the pointer to be removed from the surface, so there are no `swipestart` and `swipeend` events.

### Pinch
`pinchstart`, `pinchend`, `pinch` events. There currently are no directional events. In reality 2 directions exist here: moving the pointers towards each other or away from each other.

### Rotate
`rotatestart`, `rotateend`, `rotate` events. Like with `Pinch`, no directional events. Those could be something like `rotatecw` and `rotateccw`.

### TwoFingerPan
Same as pan: `twofingerpanstart`, `twofingerpanend`, `twofingerpan`, `twofingerpanup`, `twofingerpandown`, `twofingerpanleft`, `towfingerpanright`. No `twofingerswipe` exists currently.

**Comment by SE2Dev** [comment]
Essentially in order to resolve #28 we'd end up augmenting `HTMLElementEventMap` with ContactJS generated events to allow for strict type checking on those events.

It'd look something like this:
```ts
import { GestureEvent } from "./gestures/Gesture";

declare global {
    // Augment with ContactJS event listeners
    interface HTMLElementEventMap {
        // Tap
        "tap": GestureEvent,

        // Press
        "pressstart": GestureEvent,
        "pressend": GestureEvent,
        "press": GestureEvent,

        // Pan
        "panstart": GestureEvent,
        "panend": GestureEvent,
        "pan": GestureEvent,
        "panup": GestureEvent,
        "pandown": GestureEvent,
        "panleft": GestureEvent,
        "panright": GestureEvent,
        "swipe": GestureEvent,
        "swipeleft": GestureEvent,
        "swiperight": GestureEvent,
        "swipeup": GestureEvent,
        "swipedown": GestureEvent,

        // Pinch
        "pinchstart": GestureEvent,
        "pinchend": GestureEvent,
        "pinch": GestureEvent,

        // Rotate
        "rotatestart": GestureEvent,
        "rotateend": GestureEvent,
        "rotate": GestureEvent,
        "rotatecw": GestureEvent,
        "rotateccw": GestureEvent,

        // TwoFingerPan
        "twofingerpanstart": GestureEvent,
        "twofingerpanend": GestureEvent,
        "twofingerpan": GestureEvent,
        "twofingerpanup": GestureEvent,
        "twofingerpandown": GestureEvent,
        "twofingerpanleft": GestureEvent,
        "towfingerpanright": GestureEvent,
    }
}

export { };
```

And later used like this:
```ts
const div = document.getElementsByTagName("div")[0];

div.addEventListener("panstart", e => {
  // e is a `GestureEvent`
  console.log(e);
});
```

I dislike the idea of maintaining this list separately from the rest of the code as it would most likely result in the types becoming de-synced from the actual logic at some point, and since there's no real way for us to ensure that they'd still be in sync - a release would likely be created with invalid types.

I think that, ideally, we'd want (or maybe need?) to determine a list of all valid events for a given `Gesture` sub-type at some point in time as it would be pretty useful for automatically resolving the whole list, but I need to think about how we'd approach that more.
